helix-view/src/view.rs

```diff
@@ -66,24 +66,32 @@ pub struct View {
     pub jumps: JumpList,
     /// the last accessed file before the current one
     pub last_accessed_doc: Option<DocumentId>,
+    pub config: crate::editor::Config,
 }
 
 impl View {
-    pub fn new(doc: DocumentId) -> Self {
+    pub fn new(doc: DocumentId, config: crate::editor::Config) -> Self {
         Self {
             id: ViewId::default(),
             doc,
             offset: Position::new(0, 0),
             area: Rect::default(), // will get calculated upon inserting into tree
             jumps: JumpList::new((doc, Selection::point(0))), // TODO: use actual sel
             last_accessed_doc: None,
+            config,
         }
     }
 
     pub fn inner_area(&self) -> Rect {
-        // TODO: not ideal
-        const OFFSET: u16 = 7; // 1 diagnostic + 5 linenr + 1 gutter
-        self.area.clip_left(OFFSET).clip_bottom(1) // -1 for statusline
+        const OFFSET: usize = 1 + 1; // 1 diagnostic + 1 gutter
+        let last_line = self.offset.row + self.area.height as usize - 1; // -1 for statusline
+        let offset = match self.config.line_number {
+            LineNumber::Absolute | LineNumber::Relative => {
+                last_line.to_string().chars().count() + OFFSET
```

Comment:
I just opened a log file with a few million lines and was searching for an issue about cut-off line numbers, nice to see that it’s fixed here!
I think computing the offset is more efficient with
```rust
/// Computes the number of decimal digits needed to print a number.
fn digits10(i: usize) -> usize {
    ((i + 1) as f64).log10().ceil() as usize
}
```

Reply:
Huum I'm not sure that using float and log10 would be any fast :thinking: 

What do you think of this function?
```rust
/// Computes the number of decimal digits needed to print a number.
fn digits10(n: usize) -> usize {
    std::iter::successors(Some(n), |n| {
        let n = n / 10;
        (n != 0).then(|| n)
    }).count()
}
```

I'll run a little benchmark just to get an idea but thanks for the suggestion.
I don't think this PR will be merged any time soon however because I don't really have any time to invest in helix currently so you can take it if you want and have the time :+1:

Reply:
Here is my result with a bunch of numbers:
```
log10/0                     time:   [6.8428 ns 6.8473 ns 6.8525 ns]                            
divide/0                    time:   [1.2312 ns 1.2334 ns 1.2362 ns]                             

log10/100                   time:   [8.8163 ns 8.8380 ns 8.8643 ns]                              
divide/100                  time:   [1.6460 ns 1.6472 ns 1.6485 ns]                               

log10/1000                  time:   [9.0255 ns 9.0385 ns 9.0530 ns]                               
divide/1000                 time:   [1.9341 ns 1.9369 ns 1.9403 ns]                                

log10/4000                  time:   [9.3187 ns 9.3493 ns 9.3821 ns]                               
divide/4000                 time:   [1.8846 ns 1.8853 ns 1.8860 ns]                                

log10/10000                 time:   [8.7732 ns 8.7818 ns 8.7922 ns]                                
divide/10000                time:   [2.0882 ns 2.0900 ns 2.0922 ns]                                 

log10/50000                 time:   [9.1107 ns 9.1342 ns 9.1616 ns]                                
divide/50000                time:   [2.1692 ns 2.1741 ns 2.1798 ns]                                 

log10/100000                time:   [9.0641 ns 9.0708 ns 9.0776 ns]                                 
divide/100000               time:   [2.4120 ns 2.4160 ns 2.4204 ns]                                  

log10/500000                time:   [9.3445 ns 9.3639 ns 9.3847 ns]                                 
divide/500000               time:   [2.3472 ns 2.3540 ns 2.3627 ns]                                  

log10/18446744073709551615  time:   [7.7829 ns 7.7914 ns 7.8008 ns]
divide/18446744073709551615 time:   [10.872 ns 10.879 ns 10.887 ns]
```
`log10` is your function and `divide` is my function. The number after the `/` is which number was given to our function.

Reply:
Nice, that’s a lot faster!